rust/capture/src/utils.rs

```diff
@@ -36,3 +36,148 @@ pub fn uuid_v7() -> Uuid {
+            }
+        }
+        i += 1;
+    }
+
+    // Convert bytes back to String
+    String::from_utf8(bytes)
+}
+
+fn is_next_low_surrogate(bytes: &[u8], start: usize) -> bool {
+    let len = bytes.len();
+    if start + 6 <= len && bytes[start] == b'\\' && bytes[start + 1] == b'u' {
+        let code_point_bytes = &bytes[start + 2..start + 6];
+        if let Ok(code_point_str) = std::str::from_utf8(code_point_bytes) {
+            if let Ok(num) = u16::from_str_radix(code_point_str, 16) {
+                return (0xDC00..=0xDFFF).contains(&num);
+            }
+        }
+    }
+    false
+}
+
+#[cfg(test)]
+mod test {
+
+    #[test]
+    pub fn treplace_unpaired_high_surrogate() {
+        let json_str = r#"{"key":"\uD800"}"#.to_string();
+        let expected = r#"{"key":"\uFFFD"}"#.to_string();
+        assert_eq!(
+            super::replace_invalid_hex_escape_strings(json_str).unwrap(),
+            expected
+        );
+    }
+
+    #[test]
+    pub fn replace_unpaired_low_surrogate() {
+        let json_str = r#"{"key":"\uDC00"}"#.to_string();
+        let expected = r#"{"key":"\uFFFD"}"#.to_string();
+        assert_eq!(
+            super::replace_invalid_hex_escape_strings(json_str).unwrap(),
+            expected
+        );
+    }
+
+    #[test]
+    pub fn replace_two_unpaired_low_surrogates() {
+        let json_str = r#"{"key":"\uDC00\uDC00"}"#.to_string();
+        let expected = r#"{"key":"\uFFFD\uFFFD"}"#.to_string();
+        assert_eq!(
+            super::replace_invalid_hex_escape_strings(json_str).unwrap(),
+            expected
+        );
+    }
+
+    #[test]
+    pub fn replace_two_unpaired_high_surrogates() {
+        let json_str = r#"{"key":"\uD800\uD800"}"#.to_string();
+        let expected = r#"{"key":"\uFFFD\uFFFD"}"#.to_string();
+        assert_eq!(
+            super::replace_invalid_hex_escape_strings(json_str).unwrap(),
+            expected
+        );
+    }
+
+    #[test]
+    pub fn replace_out_of_order_low_high_surrogates() {
+        let json_str = r#"{"key":"\uDC00\uD800"}"#.to_string();
+        let expected = r#"{"key":"\uFFFD\uFFFD"}"#.to_string();
+        assert_eq!(
+            super::replace_invalid_hex_escape_strings(json_str).unwrap(),
+            expected
+        );
+    }
+
+    #[test]
+    pub fn replace_unfinished_surrogate() {
+        let json_str = r#"{"key":"\uD800\uDC0"#.to_string();
+        let expected = r#"{"key":"\uFFFD\uFFFD"#.to_string();
+        assert_eq!(
+            super::replace_invalid_hex_escape_strings(json_str).unwrap(),
+            expected
+        );
+    }
+
+    #[test]
+    pub fn test_from_bad_data() {
+        let string = include_str!("../tests/session_recording_utf_surrogate_console.json");
+        let replaced = super::replace_invalid_hex_escape_strings(string.to_string()).unwrap();
+        let _result: serde_json::Value = serde_json::from_str(&replaced).unwrap();
```

Comment:
i guess this is implicitly testing it didn't throw?
is that obvious to rusty people? or is there an explicit way?

Reply:
Yeah, sorry, should have called it out - the `unwrap`s here are "forcing" the `Result`s to give you `Ok` values - if it can't (because they're an `Err` variant), the function panics, and the test would crash, causing a failure.

In "normal" code, `unwrap` is a smell for this reason, you're basically saying "I'm purposefully not handling the error case, and my process should die if it's ever encountered" - but in tests, it's shorthand for `assert!(result.is_ok())`

A `Result` is defined as:
```rust
enum Result<T, E> {
    Ok(T),
    Err(E)
}
```

So the fact that T and E are different types means if you try to "force" a `Result::Err` to be a `Result::Ok` (or vice-versa), there's no way (again, handwaving `unsafe` away) for any implementation of `Result` to return the `T` when what it's got is an `E` - so it's only option is to kill the process (via `panic!`)

If you remember when I had you write `unwrap_err` in sync, and the compiler gave out, that was because `Result::unwrap_err` prints the `T` as part of the stack trace while panicking in the event you call it on a `Result::Ok(T)`, and the `T` we were using didn't have any implementation that could print it as a string - so the compiler literally can't compile it (this is a trick some libraries use to force you to handle their errors - if the error doesn't implement a way to turn it into a string, you can never call `Result::unwrap`, so you're forced to handle the error case _even more_ than normal)